### PR TITLE
metrics: make sure metrics work well with sharding

### PR DESCRIPTION
**Pull request by p-se**
Refers to #2355

### CODE REVIEW

**Review by manno — DISMISSED** [review]
Making the whole service template conditional.

File: dev/README.md [hunk 1]
```diff
@@ -415,12 +417,17 @@ spec:
...
-
+      shard-default: true
+      # or:
+      # shard: <shard-id>
 ```
 
-This configures Prometheus to scrape the metrics from the fleet-controller. By
-accessing the Prometheus UI, you can now see the metrics from the
-Fleet-Controller. They are all prefixed with `fleet_`, which you will need to
+This configures Prometheus to scrape the metrics from the (unshared)
```
Comment (weyfonk) [inline dev/README.md:425]: nit:
```suggestion
This configures Prometheus to scrape the metrics from the (unsharded)
```

Comment (p-se) [reply dev/README.md]: Thank you!

File: charts/fleet/templates/service.yaml [hunk 2]
```diff
@@ -1,8 +1,13 @@
-{{- if .Values.metrics.enabled }}
+{{ $shards := list "" }}
+{{ if .Values.shards }}
+{{ $shards = concat $shards .Values.shards | uniq }}
+{{ end }}
+{{ range $shards }}
+{{- if $.Values.metrics.enabled }}
```
[the remark this answers is not in the document]
Comment (p-se) [reply charts/fleet/templates/service.yaml]: Yes, absolutely right. I fixed it. Thanks!

File: e2e/assets/metrics/fleetcontroller_service.yaml [hunk 3]
```diff
@@ -2,10 +2,18 @@ apiVersion: v1
...
+    app: fleet-controller
+    env: test
 spec:
   selector:
     app: fleet-controller
-    shard-default: "true"
+    {{- if .IsDefaultShard }}
+    shard-default: "{{ .IsDefaultShard }}"
+    {{- end }}
+    {{- if ne .Shard "" }}
```
Comment (weyfonk) [inline e2e/assets/metrics/fleetcontroller_service.yaml:14]: :thinking: isn't this equivalent to `{{- if not .IsDefaultShard }}`, so to an `else` block in this context?
Do we use `IsDefaultShard` to keep a non-empty default shard as a possibility?

File: .github/workflows/nightly-ci.yml [hunk 4]
```diff
@@ -111,6 +111,7 @@ jobs:
 
           # 2. Run tests for metrics
           ginkgo e2e/metrics
+          SHARD=shard1 ginkgo e2e/metrics
```
Comment (weyfonk) [inline .github/workflows/nightly-ci.yml:114]: I find it interesting that we explicitly run tests twice here and in `e2e-ci.yml`, separating between sharded and unsharded contexts, which made me wonder: would we want to eventually expose the relevant shard (ID or empty value) in labels for metrics? How much sense could it make to users?

File: e2e/metrics/cluster_test.go [hunk 5]
```diff
@@ -87,6 +88,9 @@ func expectMetrics(name string, expectedMetrics MetricsSelector, expectExist boo
 
 var _ = Describe("Cluster Metrics", Label("cluster"), func() {
 	It("should have metrics for the local cluster resource", func() {
+		if os.Getenv("SHARD") != "" {
```
Comment (weyfonk) [inline e2e/metrics/cluster_test.go:91]: I am not following this... Why not check the `shard` variable directly? Why would this not be set for cluster tests if it is for the whole test suite?

File: e2e/assets/metrics/fleetcontroller_service.yaml (hunk 3 above)
Comment (p-se) [reply e2e/assets/metrics/fleetcontroller_service.yaml:14]: No, a non-empty default shard is no possibility. This should be simplified.

File: .github/workflows/nightly-ci.yml (hunk 4 above)
Comment (weyfonk) [reply .github/workflows/nightly-ci.yml:114]: Ah nice, this is helpful already, thanks!

**Pull request merged**